### PR TITLE
overlord/snapstate: override Snapstate.UserID in refresh if the installing user is gone

### PR DESCRIPTION
**Pull request by pedronis**
This lets override Snapstate.UserID by a refresh if the installing user has since logged out.

Do we want this behaviour? or only some explicit --reset-user option for switch and refresh? do we need a way to notify that a snap is in this state?

### CODE REVIEW

**Comment by mvo5** [comment]
The test failure:
```
FAIL: handlers_link_test.go:174: linkSnapSuite.TestDoLinkSnapSuccessUserIDAlreadySet

handlers_link_test.go:205:
    c.Check(snapst.UserID, Equals, 1)
... obtained int = 2
... expected int = 1
```
in the unit tests looks real.

**Comment by pedronis** [comment]
@mvo5 yes, the tests need adjusting, I mentioned it in the description

**Comment by pedronis** [comment]
adjusted the tests, also now based on #4398

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -607,8 +608,17 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
-		snapst.UserID = snapsup.UserID
+	if snapsup.UserID > 0 {
+		var user *auth.UserState
+		if snapst.UserID != 0 {
+			user, err = auth.User(st, snapst.UserID)
+			if err != nil && err != auth.ErrInvalidUser {
+				return err
+			}
+		}
+		if user == nil {
```
Comment (mvo5) [inline overlord/snapstate/handlers.go]: Maybe a small comment here like: `if the original user installing the snap is no longer available transfer to to user who triggered this change` (or similar :)

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -607,8 +608,20 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
+			// if the original user installing the snap is
+			// no longer available transfer to to user who
```
Comment (chipaca) [inline overlord/snapstate/handlers.go]: `to to`

File: overlord/snapstate/handlers.go [hunk 3]
```diff
@@ -607,8 +608,20 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 	// only set userID if unset in snapst and if we actually have
 	// an associated user
```
Comment (chipaca) [inline overlord/snapstate/handlers.go]: this comment probably needs updating

also: is it right to leave snapst.UserID alone if the user no longer exists but snapsup.UserID is unset?

**Pull request merged**